lib/usagereporter/teleport/aggregating/service.go

```diff
@@ -70,6 +72,93 @@ func prepareUserActivityReport(
 	return report, nil
 }
 
+// resourcePresenceReportKey returns the backend key for a resource presence report with
+// a given UUID and start time, such that reports with an earlier start time
+// will appear earlier in lexicographic ordering.
+func ResourcePresenceReportKey(reportUUID uuid.UUID, startTime time.Time) []byte {
+	return backend.Key(ResourcePresenceReportsPrefix, startTime.Format(time.RFC3339), reportUUID.String())
+}
+
+// prepareResourcePresenceReport prepares a resource presence report for storage.
+// It returns a slice of reports for case when single report is too large to fit in a single item.
+// As even a single resource kind report can be too large to fit in a single item, it may return
+// multiple reports with single resource kind report split into multiple fragments.
+func prepareResourcePresenceReports(
+	clusterName, reporterHostID []byte,
+	startTime time.Time, records []*prehogv1.ResourceKindPresenceReport,
+) ([]*prehogv1.ResourcePresenceReport, error) {
+	reports := make([]*prehogv1.ResourcePresenceReport, 0, 1) // at least one report
+
+	for len(records) > 0 {
+		reportUUID := uuid.New()
+		report := &prehogv1.ResourcePresenceReport{
+			ReportUuid:          reportUUID[:],
+			ClusterName:         clusterName,
+			ReporterHostid:      reporterHostID,
+			StartTime:           timestamppb.New(startTime),
+			ResourceKindReports: records,
+		}
+
+		if proto.Size(report) <= maxItemSize {
+			// The last report fits, so we're done.
+			reports = append(reports, report)
+			return reports, nil
+		}
+
+		// We're over the size limit, so we need to split the report and try again.
+		report.ResourceKindReports = make([]*prehogv1.ResourceKindPresenceReport, 0, len(records)-1)
+
+		emptyKindReportSize := proto.Size(&prehogv1.ResourceKindPresenceReport{
+			ResourceKind: records[0].GetResourceKind(),
+			ResourceIds:  make([]uint64, 1), // need to have non-empty slice to calculate size
+		})
+		singleItemSize := int(unsafe.Sizeof(uint64(0)))
+
+		for _, kindReport := range records {
+			if proto.Size(report)+proto.Size(kindReport) <= maxItemSize {
+				// This kind report fits, so we're done.
+				report.ResourceKindReports = append(report.ResourceKindReports, kindReport)
+				records = records[1:]
+				continue // try next kind report
+			}
+
+			freeSize := maxItemSize - proto.Size(report)
+			elementsToFit := (freeSize - emptyKindReportSize) / singleItemSize
+			if elementsToFit <= 0 {
+				// This kind report is too big to fit even if we remove all elements,
+				break // try to insert it into next report
+			}
```

Comment:
It kinda works here because a repeated `fixed64` in protobuf wire format is just packed together, but honestly I'd make the logic simpler:
- put KindNodes last in the order, as a special case
- group per-kind reports together, either skipping ones that don't fit (if you go for this we don't really need to put nodes last) or stopping when the next one doesn't fit; at this point you should have some reports ready and some kinds that don't fit even if just by themselves in a report
- for each of those kinds (likely only nodes, but maybe out there there's some cluster with 8000 databases) repeatedly split the list of resources in half every time (rather than trying to get an exact calculation right) until the report fits the limit.

Reply:
Implemented your suggestions as follows:
 - group per-kind reports together, skipping ones that don't fit
 - for each of those kinds that don't fit repeatedly split the list of resources in half every time

Rewrote test to match expected data layout: first kind is 100k records, second kind 10k, third 1k, so on.